src/tools/cephfs/JournalTool.cc

```diff
@@ -492,6 +539,243 @@ int JournalTool::journal_reset()
 }
 
 
+/**
+ * Selective offline replay which only reads out dentries and writes
+ * them to the backing store iff their version is > what is currently
+ * in the backing store.
+ *
+ * In order to write dentries to the backing store, we may create the
+ * required enclosing dirfrag objects.
+ *
+ * Test this by running scavenge on an unflushed journal, then nuking
+ * it offline, then starting an MDS and seeing that the dentries are
+ * visible.
+ *
+ * @param metablob an EMetaBlob retrieved from the journal
+ * @param dry_run if true, do no writes to RADOS
+ * @param consumed_inos output, populated with any inos inserted
+ * @returns 0 on success, else negative error code
+ */
+int JournalTool::scavenge_dentries(
+    EMetaBlob const &metablob,
+    bool const dry_run,
+    std::set<inodeno_t> *consumed_inos)
+{
+  assert(consumed_inos != NULL);
+
+  int r = 0;
+
+  // Replay fullbits (dentry+inode)
+  for (list<dirfrag_t>::const_iterator lp = metablob.lump_order.begin();
+       lp != metablob.lump_order.end(); ++lp)
+  {
+    dirfrag_t const &frag = *lp;
+    EMetaBlob::dirlump const &lump = metablob.lump_map.find(frag)->second;
+    lump._decode_bits();
+    object_t frag_oid = InodeStore::get_object_name(frag.ino, frag.frag, "");
+
+    dout(4) << "inspecting lump " << frag_oid.name << dendl;
+
+    // Check for presence of dirfrag object
+    uint64_t psize;
+    time_t pmtime;
+    r = io.stat(frag_oid.name, &psize, &pmtime);
+    if (r == -ENOENT) {
+      dout(4) << ": Frag object " << frag_oid.name << " did not exist, will create" << dendl;
+    } else if (r != 0) {
+      derr << "Unexpected error stat'ing frag " << frag_oid.name
+           << ": " << cpp_strerror(r) << dendl;
+      return r;
+    } else {
+      dout(4) << "Frag object " << frag_oid.name << " exists, will modify" << dendl;
+    }
+
+    // Update fnode in omap header of dirfrag object
+    bool write_fnode = false;
+    bufferlist old_fnode_bl;
+    r = io.omap_get_header(frag_oid.name, &old_fnode_bl);
+    if (r == -ENOENT) {
+      // Creating dirfrag from scratch
+      dout(4) << "failed to read OMAP header from directory fragment "
+        << frag_oid.name << " " << cpp_strerror(r) << dendl;
+      write_fnode = true;
+      // Note: creating the dirfrag *without* a backtrace, relying on
+      // MDS to regenerate backtraces in FSCK
```

Comment:
Don't we have code to regenerate backtraces as the dirfrag is read in without them? We should make sure we integrate properly with that if so.